As a Partner I want to edit map locations on a separate page

weliveindetail opened the pull request:
Next step for ticket https://github.com/ImmunHelden/ImmunHelden.de/issues/90

weliveindetail committed:
[Partners] Add actions to allow editing on a separate page

weliveindetail committed:
[Partners] Make function `choosePartnerId` more explicit and add notes for TODOs

weliveindetail committed:
[Partners] Minor cleanup

weliveindetail commented on new-website/src/components/location/location-table.jsx:
```diff
@@ -2,51 +2,93 @@ import React, { useState } from "react"
+        }
+        // TODO: Issue this error to sentry?
+        onError({ code: "choosePartnerId/inaccessible" })
+        throw new Error("Provided partner ID inaccessible for user")
```
Here I mimicked your approach from below. Why would we need both, calling the error handler and throwing?

c0un7-z3r0 replied:
I think the throwing was left over ... can be removed ... if you return here ... otherwise the code will continue to exceute

weliveindetail replied:
Ok, so the error hander is not a trap, but it reports the error and returns? Then we may want to rename it to `reportError`?

In this case there seems to be no useful return value and returning a dummy, I had to handle it again up-stack again. So I'd stay with the `throw` and drop the `onError`.

Just out of interest, what is your experience: Is there a significant performance penalty for throwing here (in JS)?

c0un7-z3r0 replied:
No thats fine ... but the on error function enables user notification of any errors ...

c0un7-z3r0 replied:
throw error will just crash the page if it is not  handled properly

weliveindetail commented on new-website/src/components/location/location-table.jsx:
```diff
@@ -2,51 +2,93 @@ import React, { useState } from "react"
+        if (userPartnerIds.length > 0) {
+            return userPartnerIds[0]
+        }
+        // TODO: Issue this error to sentry?
```
I assume this is one of the errors we want to see in Sentry right? How would we do this?

c0un7-z3r0 replied:
Yes 👍

weliveindetail replied:
How would we do it? Can you drop me a minimal snippet?

c0un7-z3r0 replied:
checkout the 404.jsx page

weliveindetail replied:
Done, but not sure it works. See: https://github.com/ImmunHelden/ImmunHelden.de/pull/155#discussion_r451628297

weliveindetail commented on new-website/src/components/location/location-table.jsx:
```diff
@@ -2,51 +2,93 @@ import React, { useState } from "react"
+    return async (event) => {
+        // For now, let's use a single page for add and edit
+        const docId = await newPartnerLocationDoc()
+        navigate(`/partner/location/${docId}`)
```
The page for editing is still TODO. In case the user cancels the add operation we have to remember deleting the preliminary entry. However, I did it this way on purpose, because creating the new database entry here (and not when visiting the edit page), appears to give better control over the creation of locations. The edit page will only allow to change what exists already and not create arbitrary entries (i.e. that we would otherwise need to double-check there). What do you think?

c0un7-z3r0 replied:
I would just send the user to the "add location" page and then do a error handling for collision on that page since operations on the firestore cost money 💸 so we should limit them as much as possible

weliveindetail replied:
Well, yes, they will maybe cost money in the future. Thinking again, the other problem here is, that we may stick with ghost entries when we don't get an explicit `Cancel` action (user leaves the page, connection breaks, etc.). And while we could clean them up from time to time, we always had to consider them when querying the database. So, yes, I will change that.

weliveindetail commented on new-website/src/components/location/location-table.jsx:
```diff
@@ -2,51 +2,93 @@ import React, { useState } from "react"
-            .update({ ...rest })
-            .catch(onError)
+    return async (event, rowData) => {
+        navigate(`/partner/location/${rowData.id}`)
```
How would I implement the function that handles the navigation target here? Not via `firebase.json` right?

c0un7-z3r0 replied:
Nope :D ... you can create a simple static page `src/pages/partner/location/add.jsx` and hand over the id as `location.state` property (https://www.gatsbyjs.org/docs/gatsby-link/#how-to-use-the-navigate-helper-function)

I would not try to generate for each location a separate page. no need

weliveindetail replied:
Ok, handing it over as state sounds great. Is there a way we can still have a permanent address for each location?

weliveindetail replied:
Seems like this is what I am looking for: https://www.gatsbyjs.org/docs/client-only-routes-and-user-authentication/

weliveindetail replied:
..but I don't get it to work in our case 🙁

c0un7-z3r0 commented on new-website/src/components/location/location-table.jsx:
```diff
@@ -2,51 +2,93 @@ import React, { useState } from "react"
-            .doc(id)
-            .update({ ...rest })
-            .catch(onError)
+    return async (event, rowData) => {
```
please try to avoid annonymus functions. (pain to debug)

c0un7-z3r0 commented on new-website/src/components/location/location-table.jsx:
```diff
@@ -2,51 +2,93 @@ import React, { useState } from "react"
 
 function deleteRow(onError) {
-    return async function deleteIt({ id }) {
+    return async ({ id }) => {
```
please try to avoid annonymus functions. (pain to debug)

c0un7-z3r0 commented on new-website/src/components/location/location-table.jsx:
```diff
@@ -2,51 +2,93 @@ import React, { useState } from "react"
+
+        // ShortID may collide with existing entry, retry a few times
+        for (let i=0; i<10; i++) {
+            const docId = `${partnerConfig.name}-${shortid.generate()}`
```
Why can we not just use the random id gernerated by firestore? and let firestore take care of collisions?

weliveindetail replied:
Yes, we should do that. The problem was that I encoded the pin type into the ID when hosting them in static JSON files. I did a re-import of the data and now every pin location has a type field that can be used for filtering like this:
https://github.com/ImmunHelden/ImmunHelden.de/pull/156/files#diff-1e53ef2f51bd446c876676ba83d7c888R441

Thus, I can drop the requirement for IDs to encode the type and switch to Firebase-generated ones here. IDs of imported entries still do encode their types, but from now on we can consider them to be random 😄

c0un7-z3r0 commented on new-website/src/components/location/location-table.jsx:
```diff
@@ -2,51 +2,93 @@ import React, { useState } from "react"
-            onError({ code: "requiredFieldMissing/title" })
-            throw new Error("missing title")
+function addNewRow(userPartnerIds, allPartnerConfigs, onError) {
+    const newPartnerLocationDoc = async () => {
```
please try to avoid annonymus functions. https://hackernoon.com/three-reasons-i-avoid-anonymous-js-functions-like-the-plague-7f985c27a006

c0un7-z3r0 commented on new-website/src/components/location/location-table.jsx:
```diff
@@ -74,7 +116,7 @@ export const LocationTable = ({
     const onRowUpdate = updateRow(onError)
     const onRowDelete = deleteRow(onError)
-    const onRowAdd = addNewRow(state.userAllowedPartnerIds, onError)
+    const onRowAdd = addNewRow(state.userAllowedPartnerIds, state.partnerConfigs, onError)
```
If we send the user to a new page .. we can either hand over all the infos or fetch them there from the useSession() hook

weliveindetail committed:
Revert "[Partners] Minor cleanup"

This reverts commit c57b950cf151ccede5cfcab3ea6abfd023e2dc0b.

weliveindetail committed:
[Partners] Address error handling feedback

weliveindetail committed:
[Partners] Minimal page to display location entries

weliveindetail committed:
[Partners] Minimal editing for location entries

weliveindetail committed:
[Partners] Add i18n for edit page

weliveindetail committed:
[Partners] Pass back success state to locations overview page (for review, not yet working)

weliveindetail committed:
[Partners] Make us of useSession hook on page edit location (for review, not yet working)

weliveindetail committed:
[Partners] Pass navigation-state from PartnerPage to LocationOverview and report success after editing

weliveindetail committed:
[Partners] Outsource edit-page into a separate component

weliveindetail committed:
[Partners] Second attempt for useSession on edit page (for review, still not working)

weliveindetail commented on new-website/src/components/location/edit-page.jsx:
```diff
@@ -9,7 +9,9 @@ import { LOCATION_COLLECTION } from "."
-    console.log(user, partnerConfigs, isLoading) // <- all undefined
+    const { partnerIds } = user
+
+    console.log(partnerIds, partnerConfigs, isLoading) // <- now getting `null, null, true`
```
Log output looks like this:
```
AuthContextProvider Dispatched Action
  Object
    payload: {}
    state: {user: null, isAuthenticated: false, isLoading: true}
    type: "initializing"
    __proto__: Object

AuthContextProvider Dispatched Action
  Object
    payload: {user: P, isAuthenticated: true}
    state: {user: null, isAuthenticated: false, isLoading: true}
    type: "login"
    __proto__: Object

null null true

UserContext Dispatched Action loadingUser
  Object
    payload: {}
    state: {isLoadingUser: true, isLoadingPartner: true, partnerIds: null, partnerId: null, partnerConfigs: null, …}
    type: "loadingUser"
    __proto__: Object

null null true
```

c0un7-z3r0 replied:
everything works as expected ... it is still loading the stuff ... see this screenshot 
<img width="1100" alt="image" src="https://user-images.githubusercontent.com/6809130/86638083-f4409080-bfd6-11ea-8e82-ff7600ab6808.png">

weliveindetail replied:
Yes, `edit` works but `add` doesn't

weliveindetail replied:
`add` now works. Faced a combination of issues, but the essential one here was that firestore rules prevented me from creating new database entries without passing a real `partnerId` (I did just use "test") for a while.. It took some time to figure out. See the change log for more fixes in the area.

c0un7-z3r0 commented on new-website/src/components/location/edit-form.jsx:
```diff
@@ -0,0 +1,71 @@
+  const classes = useStyles()
+  const { formatMessage } = useIntl()
+
+  const [title, setTitle] = useState(doc.title || "")
```
I have been using react-use-form just for convinience, simplicity, performance and I'm plain lazy ... see login.jsx
https://react-hook-form.com/

weliveindetail replied:
Followed your example and had a look at some of the docs, but didn't get it to work: https://github.com/weliveindetail/ImmunHelden.de/commit/82b60d3ec2932c8169553a817b62b0808c5d60d9

c0un7-z3r0 replied:
What was the error you had... I mean we can merge it without it and add it later as well

weliveindetail replied:
No error, just not getting back any values: https://github.com/weliveindetail/ImmunHelden.de/commit/82b60d3ec2932c8169553a817b62b0808c5d60d9#diff-70a4ac7d4b1a5bb8053ad94f7f0bc3b6R28

Anyway, I gave up on it nad I think it's fine.

c0un7-z3r0 commented on new-website/src/pages/partner/locations/index.jsx:
```diff
@@ -0,0 +1,53 @@
+        })
+    }
+
+    const editParam = queryString.parse(location?.search)?.edit || 'new'
```
What are all the possible values? 
Why `new`? not just null or undefined. This would include the case if `www.foo.de`, `www.foo.de?edit`, `www.foo.de?edit=`...

weliveindetail replied:
Q&D 😄 Yes, sure why not

weliveindetail replied:
On the edit page `new` is the indicator to create a new entry. It is the fallback value for `editParam` here and is used when there is no edit parameter. Rather than validating its contents here (Does the given ID exist? Is it editable for the given user?) I'd do it inside the edit page component, because there we load the document anyway.

c0un7-z3r0 commented on new-website/src/pages/partner/locations/index.jsx:
```diff
@@ -0,0 +1,53 @@
+
+    const editParam = queryString.parse(location?.search)?.edit || 'new'
+    return (
+        <ErrorBoundary>
```
This Error Boundary should not be needed unless we have an error in the base layout?

weliveindetail replied:
Fixed with https://github.com/ImmunHelden/ImmunHelden.de/pull/155/commits/f72c183134a738dd96f193fbd2296b8771c8271d

c0un7-z3r0 commented on new-website/src/components/location/location-table.jsx:
```diff
@@ -2,51 +2,91 @@ import React, { useState } from "react"
-            .update({ ...rest })
-            .catch(onError)
+function updateRow(userPartnerIds, allPartnerConfigs, reportError) {
+    return async function deleteIt(event, rowData) {
```
deleteIt was probably a copy pase error :D

weliveindetail replied:
😲 Ui, yes certainly gonna fix that

weliveindetail replied:
Done with https://github.com/ImmunHelden/ImmunHelden.de/pull/155/commits/30c54c8e8295592324670bff738d5dda7c62fc3f

c0un7-z3r0 commented on new-website/src/components/location/location-table.jsx:
```diff
@@ -2,51 +2,91 @@ import React, { useState } from "react"
+                state: { partnerId, partnerConfig, collection },
+            })
+        } catch(err) {
+            console.log(err);
```
try to avoid console.logs

weliveindetail replied:
Done

c0un7-z3r0 commented on new-website/src/components/location/location-table.jsx:
```diff
@@ -2,51 +2,91 @@ import React, { useState } from "react"
+function choosePartnerId(entryPartnerId, userPartnerIds) {
+    if (entryPartnerId) {
+        if (userPartnerIds.includes(entryPartnerId)) {
+            console.log(entryPartnerId)
```
try to avoid console.logs

weliveindetail replied:
Done

c0un7-z3r0 commented on new-website/src/pages/partner/index.jsx:
```diff
@@ -5,13 +5,13 @@ import { UserInfo } from "../../components/account/user-info"
                 <UserInfo />
                 <Invites />
-                <LocationOverview />
+                <LocationOverview state={location?.state} />
```
Please do not forward a components state reference to its children. This will cause nasty side effects

weliveindetail replied:
Ok, will change that and instead pass on the `saved` property that indicates success/failure when navigating back from the edit page.

c0un7-z3r0 commented on new-website/src/components/location/index.jsx:
```diff
@@ -61,6 +61,21 @@ export const LocationOverview = () => {
+        })
+    }
+
+    React.useEffect(() => {
```
This use effect will run at every render. Please try to add a second parameter when this useEffect should run.
https://reactjs.org/docs/hooks-reference.html#conditionally-firing-an-effect

weliveindetail replied:
Done. Used empty dependency array.

c0un7-z3r0 commented on new-website/src/components/location/index.jsx:
```diff
@@ -61,6 +61,21 @@ export const LocationOverview = () => {
+
+    React.useEffect(() => {
+        if (state?.saved == "success") {
+            state.saved = "reported" // Feels like a hack
```
why do you need to do this hack?

weliveindetail replied:
It avoids an update loop that was crashing execution eventually. IIUC it happened like this: Render Page → Effect → Trigger Alert (Re-render?) → Effect → Trigger Alert (Re-render?) → Effect → ...

Does it make sense? Likely to face a symptom of a different problem here, that can be avoided more easily.

weliveindetail replied:
This sounds like a "dependency array" I might fix it:
>  Maximum update depth exceeded. This can happen when a component calls setState inside useEffect, but useEffect either doesn't have a dependency array, or one of the dependencies changes on every render.

c0un7-z3r0 replied:
https://github.com/ImmunHelden/ImmunHelden.de/pull/155/files#r450441118

weliveindetail replied:
Done. Used empty dependency array.

weliveindetail committed:
[Partners] Avoid `useDocument` function from firestore hooks plugin and instead implement DB access manually

weliveindetail committed:
[Partners] Fix handling of the edit page result state

weliveindetail committed:
[Partners] Fix submit and drop unused parameters of edit-form

weliveindetail committed:
[Partners] Drop error boundary on edit page

weliveindetail committed:
[Partners] Simplify `add` and `update` operations in location table

weliveindetail committed:
[Partners] Fix error handling (understanding that `navigate` function does return)

I previously assumed React's `navigate` function to be a trap and not return. However, this is not the case. It does return and execute the remaining call stack..

weliveindetail committed:
[Partners] Track errors from logical inconsistencies in Sentry

weliveindetail commented on new-website/src/components/location/edit-page.jsx:
```diff
@@ -11,8 +12,15 @@ function fetchDoc(docId) {
+      Sentry.captureEvent({
+        message: "Attempt to create map entry without associated partner ID"
+      })
+    })
```
@c0un7-z3r0 When testing locally, I don't see anything popping up in our Sentry dashboard. Am I doing something wrong, or does it only work once it's deployed? (or does it have a lag?)

c0un7-z3r0 replied:
It only works on deployed versions :D We do not want to track local dev errors ;)

weliveindetail commented:
Two open issues, for details please see my comments in the unresolved conversations: So far I didn't succeed to test the Sentry event capture as well as using react-hook-form on the edit page. Otherwise, I think this is ready now.
